### PR TITLE
Move SpawnedTask from datafusion_physical_plan to new `datafusion_common_runtime` crate

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change

It seems that, we can benefit from `SpawnedTask` in more places, if we move it to under `datafusion_common`. With this, we can remove some of the linter hints `#[allow(clippy::disallowed_methods)]` in the codebase.

However, please note that with this change dependency `tokio` is introduced to the `datafusion_common` also. I am not sure if this change is desirable.

### CODE REVIEW

File: datafusion/common/Cargo.toml [hunk 1]
```diff
@@ -59,6 +59,7 @@ object_store = { workspace = true, optional = true }
 parquet = { workspace = true, optional = true, default-features = true }
 pyo3 = { version = "0.20.0", optional = true }
 sqlparser = { workspace = true }
+tokio = { workspace = true }
```
Comment (mustafasrepo) [inline datafusion/common/Cargo.toml:62]: tokio dependency is added to the datafusion_common

File: datafusion/core/src/dataframe/mod.rs [hunk 2]
```diff
@@ -2169,15 +2169,14 @@ mod tests {
     }
 
     #[tokio::test]
-    #[allow(clippy::disallowed_methods)]
```
Comment (mustafasrepo) [inline datafusion/core/src/dataframe/mod.rs:2172]: Now, this code is visible under `datafusion/core` also.

File: datafusion/sqllogictest/bin/sqllogictests.rs [hunk 3]
```diff
@@ -88,8 +90,7 @@ async fn run_tests() -> Result<()> {
     // modifying shared state like `/tmp/`)
     let errors: Vec<_> = futures::stream::iter(read_test_files(&options)?)
         .map(|test_file| {
-            #[allow(clippy::disallowed_methods)] // spawn allowed only in tests
-            tokio::task::spawn(async move {
+            SpawnedTask::spawn(async move {
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/bin/sqllogictests.rs:93]: Now this struct is visible here

File: datafusion/common/Cargo.toml (hunk 1 above)
Comment (alamb) [reply datafusion/common/Cargo.toml:62]: I also worry about this slightly, though the liklihood of someone using datafusion without tokio is likely low

I had two alternate ideas for consideration:
1.  Move to`datafusion_execution` crate (which would also then get a tokio dependence, but it is less widely used) 
2. Make a new crate to hold common execution pieces (e.g `datafusion_common_runtime` or `datafusion_common_tokio` or`datafusion_common_execution`) 

The new crate might seem like overkill initially, but I think there are several things that could probably go there to allow for better reuse. For example:
* https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-plan/src/stream.rs 
* https://github.com/apache/arrow-datafusion/blob/9aa01f35cf512e03196799c44cfecf7fe8af90e2/datafusion/physical-plan/src/common.rs#L99

**Review by alamb — COMMENTED** [review]
Thanks @mustafasrepo  -- I think this is a good idea . Like you I am worried about the tokio dependnecy in common. What do you think about making a new crate for this kind of common runtime infrastructure?

**Review by alamb — COMMENTED** [review]
Thank you @mustafasrepo 

Perhaps we can also re-export this new module in the datafusion core crate as well (can be a follow on PR). That way people can still use this code via the core datafusion crate (and don't have to add an explicit new dependnecy on `datafusion-common-runtime`?

For example, 

https://github.com/apache/arrow-datafusion/blob/b2ff249bfb918ac6697dbc92b51262a7bdbb5971/datafusion/core/src/lib.rs#L481-L483

Maybe add something like

```rust
pub mod common {
    pub use datafusion_common::*;
    pub mod runtime {
      pub use datafusion_common_runtime::*;
    }
}
```

**Review by alamb — APPROVED** [review]

**Comment by mustafasrepo** [comment]
> Thank you @mustafasrepo
> 
> Perhaps we can also re-export this new module in the datafusion core crate as well (can be a follow on PR). That way people can still use this code via the core datafusion crate (and don't have to add an explicit new dependnecy on `datafusion-common-runtime`?
> 
> For example,
> 
> https://github.com/apache/arrow-datafusion/blob/b2ff249bfb918ac6697dbc92b51262a7bdbb5971/datafusion/core/src/lib.rs#L481-L483
> 
> Maybe add something like
> 
> ```rust
> pub mod common {
>     pub use datafusion_common::*;
>     pub mod runtime {
>       pub use datafusion_common_runtime::*;
>     }
> }
> ```

Re-exported from core. Thanks for the suggestion.

**Comment by alamb** [comment]
Thanks again @mustafasrepo

**Pull request merged**